prow.sh

```diff
@@ -0,0 +1,667 @@
+#   CSI_PROW_DEPLOYMENT variable can be set in the
+#   .prow.sh of each component when there are breaking changes
+#   that require using a non-default deployment. Otherwise
+#   a deployment named "kubernetes-x.yy" is used.
+#   "none" disables the deployment of the hostpath driver.
+#
+# When no deploy script is found (nothing in `deploy` directory,
+# CSI_PROW_HOSTPATH_REPO=none), nothing gets deployed.
+configvar CSI_PROW_HOSTPATH_VERSION master "hostpath driver"
+configvar CSI_PROW_HOSTPATH_REPO https://github.com/kubernetes-csi/csi-driver-host-path "hostpath repo"
+# Ignore 'See if you can use ${variable//search/replace} instead.'
+# shellcheck disable=SC2001
+configvar CSI_PROW_DEPLOYMENT "kubernetes-$(echo "${CSI_PROW_KUBERNETES_VERSION}" | sed -e 's/\([0-9]*\)\.\([0-9]*\).*/\1.\2/')" "deployment"
+
+# If CSI_PROW_HOSTPATH_CANARY is set (typically to "canary", but also
+# "1.0-canary"), then all image versions are replaced with that
+# version tag.
+configvar CSI_PROW_HOSTPATH_CANARY "" "hostpath image"
+
+# The E2E testing can come from an arbitrary repo. The expectation is that
+# the repo supports "go test ./test/e2e -args --storage.testdriver" (https://github.com/kubernetes/kubernetes/pull/72836)
+# after setting KUBECONFIG. As a special case, if the repository is Kubernetes,
+# then `make WHAT=test/e2e/e2e.test` is called first to ensure that
+# all generated files are present.
+#
+# CSI_PROW_E2E_REPO=none disables E2E testing.
+#
+# TODO (?): support also an E2E test suite in the current repository.
+# TODO: upstream Kubernetes
+configvar CSI_PROW_E2E_REPO https://github.com/pohly/kubernetes "E2E repo"
+configvar CSI_PROW_E2E_VERSION storage-external-snapshot "E2E version"
+configvar CSI_PROW_E2E_IMPORT_PATH k8s.io/kubernetes "E2E package"
+
+# These control which tests run resp. don't run under "ginkgo -p".
+# Tests that depend on alpha features are disabled by default because
+# all tests with a feature tag are skipped.
+#
+# Note that feature tags in the test are set based on what is an alpha
+# feature in the *current* Kubernetes. When updating the test via
+# the CSI_PROW_E2E variables above and testing against an older Kubernetes
+# release, some tests might be enabled by default which don't pass
+# for that older Kubernetes version. In that case, the CSI_PROW_E2E_PARALLEL_SKIP_<k8s-version>
+# variable must be extended. <k8s-version> is just major+minor version
+# separated by underscore.
+configvar CSI_PROW_E2E_PARALLEL_FOCUS 'External.Storage.*csi-hostpath' "parallel E2E focus"
+# Raw block was an alpha feature in 1.13.
+configvar CSI_PROW_E2E_PARALLEL_SKIP_1_13 '|\[Testpattern: Dynamic PV .block volmode.\]' "additional parallel test skip for Kubernetes 1.13"
+configvar CSI_PROW_E2E_PARALLEL_SKIP "\\[Feature:|\\[Disruptive\\]$(eval echo "\${CSI_PROW_E2E_PARALLEL_SKIP_${csi_prow_kubernetes_version_suffix}}")" "parallel E2E skip"
+
+# After the parallel E2E test without alpha features, a test cluster
```

Comment:
Instead of running in the same job, would it be simpler to have 1 job for non-alpha, and 1 job for alpha?  Then they can also run in parallel, and it's easier to add more jobs in the future, like serial.

Reply:
This can be supported by allowing `CSI_PROW_E2E_PARALLEL_FOCUS` to be `none`. Then a Prow job which is only for alpha features can disable the parallel tests and vice versa.

Will add that.

Reply:
Let me add that I am not sure whether it is desirable to run these tests in parallel in different jobs.

By parallelizing, the time to completion of all tests goes down. But it wastes resources in the test cluster because the job and test cluster setup needs to be done multiple times. It's a tradeoff.

Reply:
I think it is fine to have more jobs run in parallel.  That's why prow has boskos for project lending.  Some of these serial tests takes hours.

Reply:
Not for the storage test suite. I don't think we actually have test marked up as "Serial" at the moment.

Reply:
We have two reconstruction tests (one of which is skipped for hostpath but actually it only needs to be skipped for in-tree and not csi): https://k8s-testgrid.appspot.com/sig-storage-kubernetes#gce-serial&include-filter-by-regex=csi&include-filter-by-regex=CSI.*Volumes

Reply:
Here's a job where all four combinations of stable/alpha and serial/parallel were active: https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/kubernetes-csi_csi-driver-host-path/16/pull-sig-storage-csi-driver-host-path-kubernetes-1-13/8

Most of the tests are parallel, one is serial (and fails because of the missing external pod IP), two are parallel+alpha, none are serial+alpha.

That empty test run for serial+alpha is cheap in this case, just a few seconds.

I'm now doing some post-processing of the JUnit XML files to filter out the vast numbers of unrelated tests. That makes loading in Spyglass quite noticably faster. There are also no duplicate entries for "skipped tests", so the `77/163 Tests Passed!` and `84/163 Tests Skipped.` should be accurate.